ch02.adoc

```diff
@@ -163,5 +163,76 @@ The global **`external_variables`** attribute is a blank-separated list of the n
 These variables are to be found in other files (called "external files") but CF does not provide conventions for identifying the files concerned.
 The only attribute for which CF standardises the use of external variables is **`cell_measures`**.
 
+[[groups, Section 2.7, "Groups"]]
+=== Groups
 
+Groups provide a powerful mechanism to structure data hierarchically.
+This convention does not standardize group names.
+It may be of benefit to name groups in such a way that human readers can interpret them.
+However, files that conform to this standard shall not require software to interpret or decode information from group names.
+References to out-of-group variable and dimensions shall be found by applying the scoping rules outlined below.
 
+==== Scope
+
+Any variable or dimension can be referred to, as long as it can be found with one of the following search strategies:
+
+* Search by absolute path
+* Search by relative path
+* Search by proximity
+
+These strategies are explained in detail in the following sections.
+
+If any dimension of an out-of-group variable has the same name as a dimension of the referring variable, the two must be the same dimension (i.e. they must have the same netCDF dimension ID).
+
+===== Search by absolute path
+
+A variable or dimension specified with an absolute path (i.e., with a leading slash "/") is at the indicated location relative to the root group, as in a UNIX-style file convention.
+For example, a `coordinates` attribute of `/g1/lat` refers to the `lat` variable in group `/g1`.
+
+===== Search by relative path
+
+As in a UNIX-style file convention, a variable or dimension specified with a relative path (i.e., containing a slash but not with a leading slash, e.g. `child/lat`) is at the location obtained by affixing the relative path to the absolute path of the referring attribute.
+For example, a `coordinates` attribute of `g1/lat` refers to the `lat` variable in subgroup `g1` of the current (referring) group.
+Upward path traversals from the current group are indicated with the UNIX convention.
+For example, `../g1/lat` refers to the `lat` variable in the sibling group `g1` of the current (referring) group.
+
```

Comment:
When a variable references something, let's say a coordinate, that includes an upward traversal in the hierarchical tree, then upward traversal via relative paths become useful. Many (most?) variables only reference out-of-group (OOG) objects that include an upward traversal. This is because the most commonly referenced OOG objects share dimensions with the referencing variable, and dimensions must be defined at or above the level of the referencing variable. By contrast, "downward traversal" has many fewer applications. So eliminating upward but not downward traversal would drastically curtail the use of relative path searches. Perhaps the two strongest arguments for relative vs. absolute path searches are that 1.Relative paths are less fragile that absolute because entire branches of a hierarchy can be moved between datasets without altering any of the relative paths. 2. Relative paths are shorter and thus produce more legible metadata.  As for an example, consider an ensemble of temperature data from different models all regridded to the same grid. The coordinates would naturally be stored in the parent group, and the child groups would naturally each contain a temperature variable referencing the coordinates in the parent group.

Reply:
Hello @czender 

thank you for following up on this comment. I perceive that the benefit of upward traversal is valuable.

given the section:

```
A variable or dimension specified with no path (for example, `lat`) refers to the variable or dimension of that name, if there is one, in the referring group.
If not, the ancestors of the referring group are searched for it, starting from the direct ancestor and proceeding toward the root group, until it is found.
```
Do you think there is value in having both of these mechanisms enabling upward traversal, and thus having 2 different mechanisms to support?

Would it be easier to manage if only one of these were provided?

Are there already cases of both relative paths with `../` and of references with no path requiring ancestor searching in wide use, such that both mechanisms must be supported, or is one much more common and could be targeted as the supported mechanism?

thank you
mark

Reply:
My short answer is yes. There is value in having three search methods (ancestor, relative, absolute). IMHO ancestor is the most robust and logical choice to design a dataset with. Relative search is fine and less likely to break as branches are pruned into other files. Absolute paths are more precise yet fragile than the other two. YMMV.

Regarding current practice, most hierarchical netCDF4 datasets that I have seen are from satellite retrievals, and those that attempt to be CF-compliant use absolute paths. I have not seen relative paths. I think I have seen ancestor search, yet I cannot recall which datasets that was with.

The effort required to support absolute paths is the least, since those are valid input to the netCDF API. Relative paths require marginally more effort than absolute paths. Some pathname manipulation is required before accessing the netCDF API. The effort required to support ancestor search is putting a loop around the search. Neither of those algorithms is difficult to implement. The cost of supporting relative and ancestor search must of course be weighed agains the benefits of increased flexibility and robustness.

Reply:
Hello @czender 

this makes a lot of sense, thank you

I have found an explicit reference within the NetCDF documentation regarding scope for dimensions for groups:

```
Dimensions are scoped such that they are visible to all child groups. For example, you can define a dimension in the root group, and use its dimension id when defining a variable in a sub-group.
```
https://www.unidata.ucar.edu/software/netcdf/docs/groups.html

I wonder if it is worth repeating and referencing this within the CF documentation?  It was not immediately clear to me whether the scoping visibility of parent dimensions within child groups was part of this CF proposal, or part of netCDF (which it is).

(there may also be a better reference for this than the one I provide)

This could give improved context for the scope visibility presented here.

Reply:
I have included this text in f3f4675. Does this seem clearer now?

Reply:
that's helpful, thank you